### PR TITLE
Improve pinned repos handling

### PR DESCRIPTION
**Pull request by pmiossec**
Based on PR #11790 that should be merged first.

- Add a icon to clearly see anchored repositories
- Add an option to remove pinned repositories from recent repositories (to not have duplicate repositories and save space --and user brain CPU cycles--)
- ~~Remove "Maximum number of pinned recent repositories" because if user add a repository to the pinned list, it's to see it (otherwise he will remove it from the list)~~

### Before

no "pinned" icons & repos duplicated

image

### After

Settings:

image

Pinned repos with icon:

image

With pinned repos not deduplicated (as before):
image

With pinned repos deduplicated (so less repo lines displayed) :

image

## Test methodology 

- Manual

## Test environment(s) 

- Git Extensions 33.33.33
- Build 392712e717aa06c55419142557156ff185d597a3
- Git 2.45.0.windows.1
- Microsoft Windows NT 10.0.22631.0
- .NET 8.0.5
- DPI 96dpi (no scaling)
- Portable: False
- Microsoft.WindowsDesktop.App Versions

```
    Microsoft.WindowsDesktop.App 6.0.26 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 6.0.30 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 7.0.19 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
    Microsoft.WindowsDesktop.App 8.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```

## Merge strategy

I agree that the maintainer rebase merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by mstv — CHANGES_REQUESTED** [review]
🤔 As I complained about when the regarding config form was modified, the term "pinned" may apply for the dashboard. But it does not correctly describe the behavior of the repo dropdown - at least not for the mode I use:

image

(Yes, this is how it looks. And the single column is not resizable!)

In this dropdown, the first list is just the MRU entries (sorted by usage), the second list is - more or less - the list of all repos I use - in alphabetical order.

Deduplication really needs to optional. A regarding screenshot is missing in the PR description.

The "Maximum number of pinned recent repositories" needs to be kept as I do not pin repos and do not want to do it (as well as I do not know how to pin them except using the dashboard).

Failing tests need to be adapted.

**Comment by pmiossec** [comment]
> (Yes, this is how it looks. And the single column is not resizable!)

This is due to the value 40 that you put in "Combobox minimum width" setting. I don't really know the goal of this setting and always untouched the default 0 value.

And investigating the code, that's the reason of your not readable display in #11768

**Comment by mstv** [comment]
> This is due to the value 40 that you put in "Combobox minimum width" setting.

Yes, it must have been me 1-2 years ago. But I do not remember how and why - most likely while reviewing a PR.

**Comment by pmiossec** [comment]
> 🤔 As I complained about when the regarding config form was modified, the term "pinned" may apply for the dashboard. But it does not correctly describe the behavior of the repo dropdown - at least not for the mode I use:
> 
> In this dropdown, the first list is just the MRU entries (sorted by usage), the second list is - more or less - the list of all repos I use - in alphabetical order.
> 
> Deduplication really needs to optional. A regarding screenshot of the setting is missing in the PR description.
> 
> The "Maximum number of pinned recent repositories" needs to be kept as I do not pin repos and do not want to do it (as well as I do not know how to pin them except using the dashboard).
> 
> Failing tests need to be adapted.

2nd version a lot simplified that handle all your remarks...

File: src/app/GitCommands/Settings/AppSettings.cs [hunk 1]
```diff
@@ -1690,6 +1690,12 @@ public static int RecentRepositoriesHistorySize
             set => SetInt("history size", value);
         }
 
+        public static bool HidePinnedFromRecentList
+        {
+            get => GetBool("HidePinnedFromRecentList", false);
+            set => SetBool("HidePinnedFromRecentList", value);
+        }
+
```
Comment (mstv) [inline src/app/GitCommands/Settings/AppSettings.cs:1698]: Please do it like `ShowSearchCommit`.

File: tests/app/UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs [hunk 2]
```diff
@@ -331,6 +331,7 @@ static IEnumerable<(PropertyInfo property, object defaultValue, bool isNullable,
                 yield return (properties[nameof(AppSettings.RecursiveSubmodules)], 1, false, false);
                 yield return (properties[nameof(AppSettings.ShorteningRecentRepoPathStrategy)], ShorteningRecentRepoPathStrategy.None, false, false);
                 yield return (properties[nameof(AppSettings.MaxPinnedRepositories)], 0, false, false);
+                yield return (properties[nameof(AppSettings.HidePinnedFromRecentList)], false, false, false);
```
Comment (mstv) [inline tests/app/UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:334]: ditto

**Review by mstv — COMMENTED** [review]
"number of pinned repos" is the wrong term for this dropdown.
Have a look at the screenshot of deduplication in the PR description.

**Comment by mstv** [comment]
My 2ct: Yes, "top repositories" and "pin to ..." feel much clearer to me.

**Comment by gerhardol** [comment]
"top repositories" is better (major change to rename all variables/methods if that is done though)
top/recent makes sense.

I have a minor preference for anchor vs pin and am not voting for to have an icon for top repos at all

--
I usually have a short list with 5 top repos in MRU order and 50 in alphabetic order, including the top repos as I may not look there at all.
50 is too few to keep in the big list for me, I would like to have 500. But this requires the filtering RussKie started implementing, and to limit the size of the dropdown as larger than window is behaving very weird.

File: src/app/GitCommands/UserRepositoryHistory/RecentRepoInfo.cs [hunk 3]
```diff
@@ -118,7 +125,7 @@ void AddSortedRepos(bool mostRecent, List<RecentRepoInfo> addToList)
                 addToList.AddRange(
                     from caption in orderedRepos.Keys
                     from repo in orderedRepos[caption]
-                    where !mostRecent || repo.MostRecent == mostRecent
+                    where (!mostRecent && !HideTopRepositoriesFromRecentList) || repo.MostRecent == mostRecent
```
Comment (mstv) [inline src/app/GitCommands/UserRepositoryHistory/RecentRepoInfo.cs:128]: `mostRecent` and `MostRecent` should be renamed to `TopRepo`.
```suggestion
                    where repo.TopRepo == topRepo || (!topRepo && !HideTopRepositoriesFromRecentList) 
```

File: tests/app/UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs [hunk 4]
```diff
@@ -332,6 +332,7 @@ static IEnumerable<(PropertyInfo property, object defaultValue, bool isNullable,
                 yield return (properties[nameof(AppSettings.ShorteningRecentRepoPathStrategy)], ShorteningRecentRepoPathStrategy.None, false, false);
                 yield return (properties[nameof(AppSettings.MaxTopRepositories)], 0, false, false);
                 yield return (properties[nameof(AppSettings.RecentRepositoriesHistorySize)], 30, false, false);
+                yield return (properties[nameof(AppSettings.HideTopRepositoriesFromRecentList)], false, false, false);
```
Comment (mstv) [inline tests/app/UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs:335]:
```suggestion
                yield return (properties[nameof(AppSettings.HideTopRepositoriesFromRecentList)], false, isNotNullable, isISetting);
```

**Comment by pmiossec** [comment]
@gerhardol Is my answer satisfying? or you still want me to remove the anchor icon from this PR ?

I can do it easily. I think the added value of this PR is more the "hide" setting.

**Review by mstv — APPROVED** [review]
LGTM

**Comment by gerhardol** [comment]
> @gerhardol Is my answer satisfying? or you still want me to remove the anchor icon from this PR ?

I can live with the change.
I did not intend to block this change.
(but I want the list to be sorted by path rather than what the shortening happen to call them, but I missed that feedback on the previous PR...)

**Comment by pmiossec** [comment]
> I can live with the change.

I can add an hidden setting, the added complexity will be very low...

> (but I want the list to be sorted by path rather than what the shortening happen to call them, but I missed that feedback on the previous PR...)

It seems completely counterintuitive to me: we will sort on something not displayed and the result won't appear sorted and the user won't really understand why. And I imagine that it will add some more code complexity...

I at least think that it should be discussed (in another issue) and so won't be part of this PR (so I rebased it, ready to be merged).

**Pull request merged**